File: src/qcheck_ffi.mjs
import { Ok, Error as GError } from "./gleam.mjs";

export function fail(msg) {
  throw new Error(msg)
}

export function rescue_error(f) {
  try {
    return new Ok(f());
  } catch (e) {
    if (e instanceof Error) {
      // `e` should be a string
      return new GError(e.message);
    } else {
      // rethrow the error
      throw e;
    }
  }
}

File: src/qcheck_ffi.erl
-module(qcheck_ffi).

-export([rescue_error/1]).

-spec rescue_error(fun()) -> {ok, any()} | {error, string()}.
rescue_error(F) ->
    try
        {ok, F()}
    catch
        error:#{message := Message} ->
            {error, Message}
    end.
